Fix data races in Iceberg

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix data races in Iceberg

### Changes

- `schema_id_by_data_file` should not be initialized in `getDataFiles` we don't use it later (https://github.com/ClickHouse/ClickHouse/pull/81619/files#r2155328743)

### Races

- `cached_unprunned_files_for_last_processed_snapshot` can be accessed easily from multiple queries and updated in parallel (via parallel `SELECT`)
- `relevant_snapshot` can be updated via `StorageObjectStorage::update` (basically any `SELECT` query + `system.tables` via `totalBytes`/`totalRows`)
- `IcebergMetadata::updateSnapshot` is not thread safe either

**So basically we need to protect all non-constant fields in `IcebergMetadata` and `IcebergSchemaProcessor`**

So this PR uses `shared_lock` to guard everything, but allow parallel reads (for `SELECT` and co)

Fixes: https://github.com/ClickHouse/ClickHouse/pull/81619

azat commented:
>failure	python3 ./ci/jobs/scripts/workflow_hooks/new_tests_check.py	0.99	11:48:59
>No new tests have been added

@maxknv now `pr-bugfix` does not starts CI if there were no test added?

divanik commented on src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:
```diff
@@ -595,22 +595,25 @@ std::optional<Int32> IcebergMetadata::getSchemaVersionByFileIfOutdated(String da
     {
-        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot find manifest file for data file: {}", data_path);
+        std::lock_guard lock(schema_id_by_data_file_mutex);
+        auto schema_id_it = schema_id_by_data_file.find(data_path);
```
Let's move initialisation to this scope. And make schema_id_by_data_file_initialized not atomic (we always use schema_id_by_data_file when mutex is locked so there is no necessity in additional syncronization)

azat commented:
TL;DR;

I think there are more problems, to avoid posting into multiple old PRs, I will post the questions here:

- Why do we need `schema_id_by_data_file` inside `getDataFiles`? we don't use it later (https://github.com/ClickHouse/ClickHouse/pull/81619/files#r2155328743)
- why `cached_unprunned_files_for_last_processed_snapshot` is not guarded with lock? AFAICS it can be accessed easily from multiple queries and updated in parallel
- why `relevant_snapshot` is not guarded with lock? AFAICS the snapshot can be updated from multiple queries, where `StorageObjectStorage::update` is called, i.e. totalBytes/totalRows/...
- why `IcebergMetadata::updateSnapshot` is thread safe? `update` can be called simultaneously from `totalBytes`/`totalRows` methods

What am I missing?

azat commented:
For now I'm going to add lock for the whole IcebergMetadata object, as a "hot fix", but later we should fix this properly

azat committed:
Fix data races in Iceberg

There are lots of races possible in Iceberg:
- `cached_unprunned_files_for_last_processed_snapshot` can be accessed
  easily from multiple queries and updated in parallel (via parallel
  `SELECT`)
- `relevant_snapshot` can be updated via `StorageObjectStorage::update`
  (basically any `SELECT` query + `system.tables` via
  `totalBytes`/`totalRows`)
- `IcebergMetadata::updateSnapshot` is not thread safe either

**So basically we need to protect all non-constant fields in `IcebergMetadata` and `IcebergSchemaProcessor`**

So this PR uses `shared_lock` to guard everything, but allow parallel reads (for `SELECT` and co)

Also it removes `schema_id_by_data_file` initializion from
`getDataFiles`, since it is not used later (otherwise locking will be
more costly).

v2: fix deadlock - https://pastila.nl/?0002517d/d8e89293604ddfe83765f2cbb471b56f#Fa29m24uqvhqImLMqFsmTg==

azat commented:
This has been completely rewritten to avoid any data races (at least I hope), @divanik PTAL, as discussed it uses `shared_lock` to allow parallel `SELECT`s (and other things)

azat commented:
>[Stress test (amd_debug)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=82088&sha=2fed10a0d4971250c62da5a82ef75927ff13b414&name_0=PR&name_1=Stress%20test%20%28amd_debug%29) — Killed by signal (in clickhouse-server.log)

- https://github.com/ClickHouse/ClickHouse/pull/82307

azat commented on src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:
```diff
@@ -512,8 +515,14 @@ bool IcebergMetadata::update(const ContextPtr & local_context)
     {
-        cached_unprunned_files_for_last_processed_snapshot = std::nullopt;
-        schema_id_by_data_file_initialized.store(false);
+        schema_id_by_data_file.clear();
```
It will be reinitialized anyway, since only `schema_id_by_data_file_initialized` is checked, this is just to avoid storing stale data.

azat commented:
>This branch has conflicts that must be resolved
>src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp
>src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.h

Sigh.

We need to merge this ASAP, since anything added on top has the same bugs and makes it harder to backport this fix

azat commented:
>[Integration tests (asan, old analyzer, 1/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=82088&sha=3981dd5952e7757cc06d098c8b14e3e0f36b7638&name_0=PR&name_1=Integration%20tests%20%28asan%2C%20old%20analyzer%2C%201%2F6%29) — fail: 6, passed: 567

- test_database_delta - https://github.com/ClickHouse/ClickHouse/issues/79109

>[Integration tests (asan, old analyzer, 4/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=82088&sha=3981dd5952e7757cc06d098c8b14e3e0f36b7638&name_0=PR&name_1=Integration%20tests%20%28asan%2C%20old%20analyzer%2C%204%2F6%29) — Job timeout expired, fail: 0, passed: 600

Did not run any iceberg tests, but need to find the slow test
